### PR TITLE
style: function call wrapping

### PR DESCRIPTION
**Pull request by RaduBerinde**
Updates #9678.

Bikeshed away! Apart from having a consistent style, my only preference is for something that can be automatically enforced via `crlfmt` (if possible).

CC @paperstreet @dt @tamird @petermattis @bdarnell

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9687)

### CODE REVIEW

**Comment by paperstreet** [comment]
Can you include an example of a function call that takes a few arguments and then a (long) inline closure? Something like this line: https://github.com/cockroachdb/cockroach/blob/96a61d731c50bccfffcbb1e55786600f346a281a/internal/client/client_test.go#L868

**Comment by RaduBerinde** [comment]
Done.

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTBNqXI8iRDU1KX851b:bnfp4nl)*

**Comment by RaduBerinde** [comment]
@petermattis, @spencerkimball are you ok with this?

**Comment by tschottdorf** [comment]
Make sure to send a PSA when this goes in.

**Comment by andreimatei** [comment]
What's wrong with
```
  someFunc(arg1, ..., argn,
    argn+1, ... arg)
```
?

The reason we don't allow this in function declarations is that there the first line of the function is indented, and it would be indented at the same level as `argn+1`, and it'd be confusing. But not so here.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTGMWJ01G8QlN8_0ExX:bdf06ax)*

**Comment by petermattis** [comment]
I'm mildly anxious about imposing a style for function call syntax that isn't enforced by `go fmt`. `crlfmt` is nice, but it imposes a burden on new developers. On the other hand, we're the primary developers that we need to be optimizing for. Meh, I don't have strong feelings either way.

**Comment by tamird** [comment]
@andreimatei the syntax you've shown is confusing for a different reason; it produces surprising indents without a closing delimiter:
```
if err := someFunc(arg1, ..., argn,
  argn+1, ... arg); err != nil {
  panic(err)
}
```

That's pretty ugly and confusing.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTGOlSuUB6jqTaIv-qG:bp57kmi)*

**Comment by RaduBerinde** [comment]
It also looks pretty sloppy with arguments starting on different positions on different lines IMO. If we could align `argn+1` under `arg1` it would be a different story but that's not an option.

@petermattis If we don't want to impose it, we can document it as "encouraged". I think there is still value to recommending a single format and trying to stick with it if possible.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTGOOPNaBVMD76Zf2J1:bikh3od)*

**Comment by petermattis** [comment]
"encouraged" seems better to start with. 

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTGzE60gPQE0Yx5VjuL:b-47ap8c)*

**Comment by tamird** [comment]
I think enforced is better to start with. Let's not add more reviewer
burden.

**Comment by knz** [comment]
"If we could align argn+1 under arg1 it would be a different story but that's not an option." -> why is it not an option? This would have my preference, clearly.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTI71o8U-GqzZMMjPKU:b-307goq)*

**Comment by tamird** [comment]
Because gofmt disagrees.

**Comment by andreimatei** [comment]
So on this call:
```
  someFunc(arg1, ..., argn,
    argn+1, ... arg)
```
@tamird 's example is indeed not ideal, so if you guys feel strongly, let's disallow that specific case (when the function call comes right before a new block). But I wouldn't go further than this and always disallow it. I don't think it hurts readability for simple calls. And looking "sloppy"... well... looks fine to me :) gofmt is already prescriptive enough, why do we need to go so much further?

A separate issue: I noticed that #8823 recently merged by @jordanlewis introduced this check failure:
```
-func (s *Store) maybeMergeTimestampCaches(
-   subsumingRng *Replica,
-   subsumedRng *Replica,
-) error {
+func (s *Store) maybeMergeTimestampCaches(subsumingRng *Replica, subsumedRng *Replica) error {
```
So it's forcing me to fill the line up to 100 chars. Was this intended? The style guide says to wrap at 100, but I think the intention was for that to be an upper bound. Can we relax this? I think it's way too prescriptive.


On a meta note, was there something specific that prompted this agitation? :) Why do we care about `crlfmt` enforcing more rules?

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTI5wFd_TbZFVDN3tK3:bo4v5je)*

**Comment by knz** [comment]
I'm a later comer to this party, but I just spent some time carefully studying the evolution of this discussion back from #8823 through #9678 to here. At a very high level it looks to me that someone was playing around with crlfmt, came up with some random idea, and then we're trying to accommodate a team of 15+ engineers around that idea without reflecting on why we're even spending this time + effort on this right now.

So really I have a couple serious questions:

1) what is the problem that this change is trying to solve?

2) what is the reason for addressing this problem now, as opposed to later?

3) do you plan to come to a conclusion during code yellow (this is not stability related and all efforts should be, by definition of code yellow, be subordinate to stability) and while Ben is on vacation?

4) what is the rationale for automatically de-wrapping a function? (I could perhaps be ok with a fmt utility that automatically wraps 1 line to many, or a linter that asks me to do the same, but I absolutely do not understand why we would want to un-wrap. If it's wrapped, there's intent that's gone into it by the developer.)

5) there is an example *output* from the linter from #8823 that hurts my eyes:
```
func updateStatsOnPut(
	key roachpb.Key,
	origMetaKeySize, origMetaValSize,
	metaKeySize, metaValSize int64,
	orig, meta *enginepb.MVCCMetadata,
) enginepb.MVCCStats {
```
So apparently we're not even discussing putting different arguments on different lines, but rather some hybrid where arguments of the same type may be on the same line. What is the rationale for accepting this and also meanwhile refusing to put arguments of different types on the same line?


In addition to these questions there's a general request: please let's not go down the route of asking for a whitespace style without a utility that enforces this style *automatically*. `gofmt` exists for a reason, and this entire discussion suggests we're not taking its lesson seriously.  If we go down this route, please also fork `gofmt` or create a drop-in replacement and make it do The Right Thing:tm:!

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTIBP4FIKOCU5u8bgm4:b-50j6n)*

**Comment by tamird** [comment]
On Wed, Oct 5, 2016 at 1:07 AM, kena <notifications@github.com> wrote:

> 1) what is the problem that this change is trying to solve?
>
This is trying to standardize on a format for function calls which is under
specified by `gofmt`, as a precursor for automation in `crlfmt`.

> 2) what is the reason for addressing this problem now, as opposed to later?
>
When is 'later'?

> 3) do you plan to come to a conclusion during code yellow (this is not
> stability related and all efforts should be, by definition of code yellow,
> be subordinate to stability) and while Ben is on vacation?
>
All efforts are subordinate to stability? I'm not sure what you mean, since
there is plenty of ongoing work which is not stability related.

> 4) what is the rationale for automatically de-wrapping a function? (I
> could perhaps be ok with a fmt utility that automatically wraps 1 line to
> many, or a linter that asks me to do the same, but I absolutely do not
> understand why we would want to un-wrap. If it's wrapped, there's intent
> that's gone into it by the developer.)
>
You're talking about #8823 now, right? I believe this is done just because
it was the easiest thing to do in `crlfmt`. Personally, I prefer this
behaviour since it removes some useless burden of choice, but that
discussion can be had over in the `crlfmt` repo since the rule is not fully
specified in the style guide.

> 5) there is an example *output* from the linter from #8823
> <https://github.com/cockroachdb/cockroach/pull/8823> that hurts my eyes:
>
> func updateStatsOnPut(
>     key roachpb.Key,
>     origMetaKeySize, origMetaValSize,
>     metaKeySize, metaValSize int64,
>     orig, meta *enginepb.MVCCMetadata,
> ) enginepb.MVCCStats {
>
> So apparently we're not even discussing putting different arguments on
> different lines, but rather some hybrid where arguments of the same type
> may be on the same line. What is the rationale for accepting this and also
> meanwhile refusing to put arguments of different types on the same line?
>
This looks like a bug in `crlfmt`; I'd have expected all the `int64`s to be
put on the same line based on my understanding of the mechanics, but
@jordanlewis can give you more information here.

> In addition to these questions there's a general request: please let's not
> go down the route of asking for a whitespace style without a utility that
> enforces this style *automatically*. gofmt exists for a reason, and this
> entire discussion suggests we're not taking its lesson seriously. If we go
> down this route, please also fork gofmt or create a drop-in replacement
> and make it do The Right Thing™️!
>
This issue's description already explicitly states that the intention is to
automate this in `crlfmt`, so it's not clear to me what it is you're asking
for.

**Comment by RaduBerinde** [comment]
@andreimatei, @knz - why are we discussing #8823 here? There was already a discussion in that PR (and it was open for a very long time) and you can still post there or file an issue if you want to follow up. I made pretty much the same comment you guys are making, but I agreed with others who pointed out that the benefit of having a single, automatically enforced style outweighs the one-off case when you would have wanted to format it differently. The benefit applies to both writing code (you don't have to think about it) and reviewing (eliminates "you should wrap it this way" back and forth).

@andreimatei: I understand you prefer your style. Others prefer other styles. Is there any particular objective advantage to your style vs the proposed one? We already have a significant disadvantage in that we would need a separate set of rules for the `if x := funcCall(..)` case. Or are you proposing that we don't prescribe any style and let anything go (like right now)? There is a big advantage to having consistency, enough that consistency itself is more important than the particular chosen style. You get used to whatever style is common and what doesn't fit becomes distracting.

We can wait until Ben is back. But please let's stop with the "it's not stability" stuff (which sounds more like "i don't like it so let's not do it because code-yellow"; @knz how many of your recent PRs were strictly stability related?)

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTJT3nMJBevS95baqCh:bqvjzyt)*

**Comment by knz** [comment]
My comment about code yellow is that this PR is basically asking everyone in the team to pause and think about this, perhaps not long but still a little bit. Our other PRs are much more contained and do not need all-hands participation.

**Comment by RaduBerinde** [comment]
Filed https://github.com/cockroachdb/crlfmt/issues/4 for the instance you pointed out above. Note that this is due to `crlfmt` incorrectly preserving the newline that already existed between `origMetaKeySize, origMetaValSize` and `metaKeySize, metaValSize`, so it's not a problem that interferes with writing new code.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTJp31BIZQa5RLqews7:b-w157oy)*

**Comment by andreimatei** [comment]
I was discussing #8823 here because I thought it's all related and we're trying to retroactively legislate what the tool already does. Seemed fishy to me that the tool started having opinions about formatting at the same time this style update is being discussed (hmmm but the tool is called crl*fmt*... what does this tool do exactly?). I'll take my unwrapping objections there.

But in principle, yeah, I think my preference would be that we don't prescribe any more purely esthetic rules. We already live under the most strict formatting rules out there because of gofmt. Now we're talking about mandating things that were too much even for gofmt to standardize. Do we really need more rules? I like consistency too (for readability), but aren't we consistent enough already? Do you find our code unreadable? I was programming happily, minding my own business, and suddenly my code isn't good any more.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTJygIEXv2ttsFPdySd:b-upsdtm)*

**Comment by tamird** [comment]
On Wed, Oct 5, 2016 at 9:28 AM, Andrei Matei <notifications@github.com>
wrote:

> But in principle, yeah, I think my preference would be that we don't
> prescribe any more purely esthetic rules. We already live under the most
> strict formatting rules out there because of gofmt. Now we're talking about
> mandating things that were too much even for gofmt to standardize. Do we
> really need more rules? I like consistency too (for readability), but
> aren't we consistent enough already? Do you find our code unreadable? I was
> programming happily, minding my own business, and suddenly my code isn't
> good any more.


Didn't @RaduBerinde sufficiently address this?

@andreimatei: I understand you prefer your style. Others prefer other
> styles. Is there any particular objective advantage to your style vs the
> proposed one? We already have a significant disadvantage in that we would
> need a separate set of rules for the if x := funcCall(..) case. Or are you
> proposing that we don't prescribe any style and let anything go (like right
> now)? There is a big advantage to having consistency, enough that
> consistency itself is more important than the particular chosen style. You
> get used to whatever style is common and what doesn't fit becomes
> distracting.


TL;DR why are strict formatting rules bad? I can understand wanting to
avoid the manual work of reformatting, but if the tool is there to do it
for you (and I think the intention is that `crlfmt` will wrap
`goimports`/`gofmt` so you can run it on save), what's the big objection?
ISTM that you just want to "do it your way" rather than stick to any
consistent style at all, and that doesn't seem like a reasonable request to
me.

**Comment by knz** [comment]
I took some more time to reflect on this and I am settling on the
following opinions:

1) a strict(er) guide about how to format function argument/parameter
lists would probably help me; I recall enough instances where I
hesitated that I would welcome the time save. However

2) I am strongly against implementing any new formatting rule by means
of a check linter without previously publishing and announcing a drop-in
replacement to gofmt to automate the formatting for me [with a caveat,
see next point]. It's one thing to guide me when I hesitate, but another
entirely to punish me when I didn't place whitespace in the right places
while I am really trying to focus on something objectively more
important. Conversely, if there's an auto-formatter already and it
integrates in my editor, I do not care in any way whatsoever and the
following points become moot.

The following two points are provided for context, perhaps they are not
applicable since I now understand the plan is to make crlfmt ready
before the new rules take effect. If my understanding is incorrect, read on.

3) I may be willing to consider applying a new rule manually (i.e. pay
cognitive bandwidth for it) once I am faced with compelling evidence
that there will be benefits down the line. In this particular case, this
is to be achieved by pulling code examples for our projects for which
the current formatting is detrimental (in any dimension that I value)
and for which the new rule, regardless of the presence of an
auto-formatter, makes the code more readable. Unfortunately, I have not
seen such evidence so far, whereas courtesy suggests the onus to produce
such evidence this still lies on the person(s) proposing the change.

4) I am not happy being told merely "strict formatting is good, if
you're a good programmer you should recognize this". This reeks of
"privacy is bad, if you don't have anything to hide, you shouldn't care
anyway" and "we are implementing these security measures for your own
protection".  In general, to me *new rules that are not motivated by a
thorough trade-off or cost/benefit analysis with both retrospectives
into the past and projections in the future are INHERENTLY bad.* They
are bad because they are a direct instance of the broken window fallacy.
(NB: this is a much older opinion of mine, I'm willing to listen to
arguments to sway it, but it's been formed over the years so the
challenge is greater)

-- 
Raphael 'kena' Poss

**Comment by tschottdorf** [comment]
> 2) I am strongly against implementing any new formatting rule by means
of a check linter without previously publishing and announcing a drop-in
replacement to gofmt to automate the formatting for me

I agree, either do it for me (which I would like very much) or leave my code be (that said, isn't crlformat a good drop-in formatter? The only problem is that it's nonstandard and so IDE integration isn't automagical).

**Comment by RaduBerinde** [comment]
@andreimatei: the tool does not currently format function calls. There is no "retroactive legislation". The fact that the other PR was able to go in (after the branches merged) triggered this discussion, so that would explain the "fishiness".

@knz, @tschottdorf: yes, this PR does not institute any linter checks. It is merely about trying to come up with a recommendation, with the intention of making it automatic through `crlfmt` (this should be clear from the commit message and the issue). I think we all agree that a linter check without having the tool wired up to the editor would be annoying.  Note that the converse is also true: having the tool without a linter check allows people not using the tool to submit code which then results in unwanted diffs being generated automatically.

**Comment by tamird** [comment]
Can we all agree that having a consistent style is good? If so, let's refocus the discussion to what that style that should be. The current proposal LGTM.

**Comment by knz** [comment]
LGTM too

**Comment by andreimatei** [comment]
Tamir, right now we do not *all* agree that "having a consistent style" is good, if having a consistent style means enacting more rules than the ones we already have. At least not without more convincing. In other words, I think we're consistent enough already.
I, for one, already said I don't agree. Raphael said he doesn't agree in his 4) above, although he somehow seems to have changed his stance. Ben once said in a review about a change to the wrapping of a function call "before someone decided to get prescriptive about this stuff in the style guide", which I took to be a snarky comment which I think means he wouldn't agree either.

As with everything else, there are tradeoffs. The onus to prove that a change is required/beneficial should fall on the instigator. But still, I'll give some arguments against introducing more rules, and I'll let y'all tell me what are we gaining exactly. "a consistent style" is not good enough for me personally. I can read the code currently just fine, so I don't think we gain any readability.

1) An impossibility argument: if we don't mandate "unwrapping", so I'm still free to wrap wherever I see fit below 100 chars (and I sure hope we're not crossing that line too), then there's no perfect consistency anyway. Also consistency is not something we value above everything else. We don't force consistency to the extreme in other aspects of coding, why should we do it in formatting? E.g. we don't always `Unlock()` in a `defer` - sometimes we do, sometimes we don't. Depends on the context. Let me decide what is best, depending on the context. This doesn't preclude us from putting some non-binding examples in the style guide to serve as guidance for new people that are hesitating.
2) A libertarian argument: whatever style we do adopt, it's probably going to outlaw some lines I'll want to write at some point, giving me grief and so driving down employee morale. In particular, the current concrete proposal would always bother me. There's gotta be a compelling reason for that.
3) A deference to authority argument - I don't think Ben would approve. In more seriousness, this is a democratic point - you want to do superficial, but wide-reaching changes and you're not consulting "the people" (all of us) that will be affected. As opposed to more serious changes that should be driven by experts, this kind of style changes I think would be reasonable to enact through large majorities.
4) A tax code argument - reading a lot of rules costs time and cognitive... something. A small style guide has an advantage that it's easy to grok and harder to decay.
5) A Murphy's Law argument - rules tend to have bugs and unintended consequences. They'll need to be maintained and updated and they'll get exceptions and loopholes. And all these require precious time every step of the way (as is the case right now). 
6) A software development argument - the tool that we'd use to enforce these rules would have bugs (seems like that's already the case, 2 bugs I think were mentioned just in this thread), will need maintenance, everybody will need to be aware of it, it'll take `make check` time, etc.
7) A git argument - every new rule will introduce otherwise unnecessary changes to the codebase that will break `blame`. In particular, the current proposal, with the closing paren on a new line, would change pretty much every wrapped call in our code, as far as I can see by randomly looking at two file.
8) A conservative argument - gofmt is already very restrictive. Let me whatever minuscule freedom I still have to express myself in the code as long as it doesn't hurt readability.
8) A range against the machine argument.

**Comment by RaduBerinde** [comment]
@andreimatei I already said we can wait for Ben, so let him speak for himself. All the arguments you wrote can be applied for the function definition wrapping as well (and indeed `gofmt` itself) and yet most of us agree that setting a guiding rule, and (recently) enforcing and automatically formatting according to that rule has been a positive change.

**Comment by andreimatei** [comment]
I was using Ben as an anecdote to show that it's not as obvious as Tamir was implying that everybody agrees that "consistency" is the wholly grail. I wasn't necessarily even suggesting that we need to wait for him to say anything. Mostly that point was that the masses aren't being properly consulted. Let's not let this waiting hijack the conversation.

I still haven't really heard why any change from the status quo is desirable. Do you really think this change improves readability? Maybe you can convince me, I'm not entrenched enough in my position already that I won't see reason.

Not all my arguments apply to `gofmt`. And of the ones that would in principle apply, some don't apply any more when you consider that `gofmt` is effectively frozen. It will not get any *new* rules.
Also they apply to different degrees to the change about function definition formatting. I was mostly following that style even before it became a rule, and life was good even when reading other people's code. I don't think it was the most necessary change in the world.

But the whole point is that there are diminishing returns and escalating costs. Just because one rule is good doesn't mean the next 100 are also good. For example, one quantitative difference is that function calls are more numerous than function definitions. And also function calls are more embedded in surrounding code, so more nuance and context might be useful when deciding how to format each one. Whereas function definitions always start an isolated block. And the function definition always ran into the issue of confusing indentation of the next block, whereas function calls don't usually have that problem.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTKkpDaWE0vpa9xI9iL:b-5s10vq)*

**Comment by tschottdorf** [comment]
The proposal LGTM too but until it is automatically formatted for me I consider these non-binding suggestions and will look unfavorably upon those that nit me on their grounds.

**Comment by tamird** [comment]
+1 to @tschottdorf. I'm in favor of this change so that it can be used to
refine crlfmt. I think we should strive to have tooling take care of all
formatting.

**Comment by RaduBerinde** [comment]
> I still haven't really heard why any change from the status quo is desirable. Do you really think this change improves readability? Maybe you can convince me, I'm not entrenched enough in my position already that I won't see reason.

1) Absolutely, having a consistent style does improve readability. Many times when reading code and finding something that is unusually wrapped (unusually = not the way I'd normally it), I find that distracting. Of course it is relatively minor, but it is there.

2) Having an agreed-upon way to format saves time when writing code. I remember many occasions when I went back and forth trying out different ways of wrapping (at least until @tamird suggested the format in this PR). I don't think I am the only one (e.g. see @knz's number 1 above). 

3) Having a consistent style preempts unnecessary back and forth during reviews. We're used to different formats and something that looks fine to one person might look weird to another. You mentioned morale: the impact on morale of a fixed format (with a tool that does the formatting for you) is significantly less than the morale impact of occasional bickering during code reviews. Moreover the former happens once (when you get used to the new tool), the latter is ongoing.

4) I don't agree that the returns are diminishing. The more call sites, the better the returns (well, for those who there are some returns).


I want to ask you to honestly answer the following question to yourself - if the proposed format was the one you want, would you still be arguing against having a fixed format?  Because the first two comments in this PR were NOT that we shouldn't have a style, they were more about using a different style..

**Comment by RaduBerinde** [comment]
Anyway, I think everyone in the thread agrees that having non-binding guidance is ok. 

I added a note mentioning that these are suggestions and should not be demanded in code reviews. Let's leave it at this for a while and reopen the discussion of whether we want to enforce it AND have `crlfmt` do it automatically later (perhaps we will all get more used to this format and there will be fewer reasons to disagree). Sounds good?

**Comment by arjunravinarayan** [comment]
+1 on "'encouraged until crlfmt automates this', upon which this encouragement can change to 'mandatory'".

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTLUFRypRjEC2ept9ws:btb3fs8)*

**Comment by andreimatei** [comment]
@raduberinde of course that if you legislate exactly the style I prefer, I would complain less :) If everybody would always agree with me, I'd be very happy (although I would still like to avoid committing to being myself consistent every time, cause maybe I want different formatting depending on arbitrary context) . But the point is I can't force *others* to agree with me. Well, maybe I could have if I would have started this PR :)

I'll make one more point and then shut up - if the value of a mandated style is that it saves you from thinking and hesitating, then you can still have a tool that formats in whatever way and just use it yourself, without making it part of the guide.

Anyway, if you really were confused by function calls in the past and the parentheses would have made you happy, and if no more people speak up against this, I'll shut up too. 

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTLmLCg3qOdhW9o7jr-:b5zyi2y)*

**Comment by tamird** [comment]
> I'll make one more point and then shut up - if the value of a mandated style is that it saves you from thinking and hesitating, then you can still have a tool that formats in whatever way and just use it yourself, without making it part of the guide.

@RaduBerinde addressed this as well in https://github.com/cockroachdb/cockroach/pull/9687#issuecomment-251690705:

> Note that the converse is also true: having the tool without a linter check allows people not using the tool to submit code which then results in unwanted diffs being generated automatically.

**Comment by RaduBerinde** [comment]
^ Yeah, lol, that would be an operational nightmare.

> If everybody would always agree with me, I'd be very happy [..]

So then you don't *really* feel strongly about your arguments 4, 5, 6, 7, and 8 :)

**Comment by andreimatei** [comment]
> Note that the converse is also true: having the tool without a linter check allows people not using the tool to submit code which then results in unwanted diffs being generated automatically.

Well the tool could just be applied on the lines you edit, like the automatic wrapping for comments works in Vim (and all other editors, I assume).

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTMI0fK-EVNFno1Q16Q:b-51deee)*

**Comment by RaduBerinde** [comment]
You must realize how complicated it would be to get something like that to work..

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTML9dSFP1A_azH4H7D:bsbpy2)*

**Comment by andreimatei** [comment]
Not really... What I had in mind is that you select a few lines and press something, then the tool runs on the whole file, produces a diff and only applies the part of the diff that intersects with the selection.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9687#-:-KTMNhypQolHP9lTHgtv:b-qcm7ik)*

**Comment by bdarnell** [comment]
I agree with @andreimatei; consistency (especially in whitespace) is overrated. I don't object too strongly as long as the rule can be automated, although I would argue that automating enforcement of this rule would be a poor use of someone's time. 

I think adding a rule to the style guide that specifically declares itself "non-binding" is silly. I think that except for rules that are enforced by tools, the *entire* style guide should be considered equally "non-binding" and we shouldn't be so picky about style guide conformance in code reviews. The goal of the style guide should be to stop churn, not create it: it ensures that to the extent that we do make style-related changes they're moving in a consistent direction.

The proposed change looks OK to me aesthetically, but it differs from the way the codebase is already formatted - search for `})` to see how often we have inline functions without putting the closing paren on its own line.

**Comment by RaduBerinde** [comment]
Thanks for the input Ben.

I was also going to write that, after sleeping on this and stepping back a bit, the negatives of some people being annoyed because of an arbitrary rule (assuming we automate the formatting in the future) probably outweigh the benefits (it's only cosmetic after all, and there's at least one person who would be annoyed, at least for a while).

I'm going to close this. Sorry to waste people's time and energy.
